mlir/test/Examples/nvgpu/tools/nvdsl.py

```diff
@@ -0,0 +1,415 @@
+            )
+        else:
+            nvgpu.mbarrier_arrive(self.mbar_group_op, self.id_op, predicate=predicate)
+
+    def try_wait(self, phase: bool = False, ticks: int = 10000000):
+        ticks_op = const(ticks)
+        phase_op = const(phase, T.bool())
+        nvgpu.MBarrierTryWaitParityOp(
+            self.mbar_group_op,
+            phase_op,
+            ticks_op,
+            mbarId=self.id_op,
+        )
+
+
+class TMA:
+    """A class that builds a TMA descriptor."""
+
+    def __init__(
+        self,
+        shape,
+        memref_ty,
+        swizzle=nvgpu.TensorMapSwizzleKind.SWIZZLE_NONE,
+        l2promo=nvgpu.TensorMapL2PromoKind.L2PROMO_NONE,
+        oob=nvgpu.TensorMapOOBKind.OOB_ZERO,
+        interleave=nvgpu.TensorMapInterleaveKind.INTERLEAVE_NONE,
+    ):
+        self.swizzle = swizzle  # mlir.nvgpu.TensorMapSwizzleKind
+        self.l2promo = l2promo  # mlir.nvgpu.TensorMapL2PromoKind
+        self.oob = oob  # mlir.nvgpu.TensorMapOOBKind
+        self.interleave = interleave  # mlir.nvgpu.TensorMapInterleaveKind
+        self.shape = shape
+        self.memref_ty = memref_ty  # MemRefType
+        self.lastDim = 64
+        self.requiredLoad = 1
+        self.tma_shape = shape
+        self.tma_memref = ir.MemRefType.get(shape, memref_ty.element_type)
+
+    @property
+    def tensormap_descriptor_ty(self):
+        """Returns a tensormap descriptor type."""
+        memref_str = f"memref<{self.tma_shape[0]}x{self.tma_shape[1]}x{self.memref_ty.element_type}, 3>"
+        parse_str = f"!nvgpu.tensormap.descriptor<tensor = {memref_str},\
+                                              swizzle = {self.swizzle},\
+                                              l2promo = {self.l2promo},\
+                                              oob = {self.oob},\
+                                              interleave = {self.interleave}>"
+
+        return ir.Type.parse(parse_str)
+
+    def create_descriptor(self, device_ptr):
+        tma_descriptor_ty = self.tensormap_descriptor_ty
+        device_unranked_memref = memref.CastOp(
+            ir.UnrankedMemRefType.get(
+                self.memref_ty.element_type, self.memref_ty.memory_space
+            ),
+            device_ptr,
+        )
+        self.tma_descriptor = nvgpu.TmaCreateDescriptorOp(
+            tma_descriptor_ty, device_unranked_memref, map(const, self.tma_shape)
+        )
+        return self.tma_descriptor.result
+
+    def prefetch(self, predicate=None):
+        nvgpu.tma_prefetch_descriptor(self.tma_descriptor, predicate=predicate)
+
+    def load(self, dest, mbarrier: Mbarriers, coords=[0, 0], predicate=None):
+        coord_ops = [const(c) for c in coords]
+        nvgpu.TmaAsyncLoadOp(
+            dest,
+            mbarrier.mbar_group_op,
+            self.tma_descriptor,
+            coordinates=coord_ops,
+            mbarId=mbarrier.id_op,
+            predicate=predicate,
+        )
+
+
+class MatrixAccumulator:
```

Comment:
I would call it is `WgmmaAccumulators` or `WarpgroupMmaAccumulators` as this calls is not for every accumulators but specialized for Hopper Tensor Core accumulators.

Reply:
Absolutely, I'm committed to fixing the first issue. 

Regarding the second one, I currently lack an MLIR op for it. However, I can implement it later. Nevertheless, even with this implementation, it's important to acknowledge that we won't be able to obtain every `get_accum_at(row, col) -> vector<1xty>` due to the fragmentation of the accumulator across the warpgroup [see fragmented matrix-d here](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#wgmma-64n16-d).

Reply:
This probably not in scope of this PR, but I am digging deep into `nvgpu` and thus making a note here. 

Looking at `nvgpu.mma.sync` 

```
 %accum = nvgpu.mma.sync (%a, %b, %initial_accum) {mmaShape = [16, 8, 16]} : (vector<4x2xf16>, vector<2x2xf16>, vector<2x2xf32>) -> vector<2x2xf32>
```

The accumulators are fragmented within the warp, but `nvgpu.mma.sync` op  has the shape and elements per thread. This allows each thread for the:
1.  use `vector.extract` to access the element and apply element operation.
2. indices of the shape from `%accum : vector<2x2xf32>` can decomposed into row, col within the accumulator fragment of `16x8` within the warp-instruction.

On similar lines `nvgpu.warpgroup.async` %accum registers should be `vector<N/8x2x2xf32>`; where the `WgmmaInstructionShapeN` to achieve 1. and 2. in `nvgpu`and/or its Python bindings.